File: packages/python-sdk/trackmcp/privacy.py
from __future__ import annotations

import base64
import binascii
import json
import re
from typing import Any, Dict, Iterable, Optional, Set
from urllib.parse import parse_qsl, urlparse


DEFAULT_MAX_PAYLOAD_BYTES = 32 * 1024
DEFAULT_MAX_PAYLOAD_DEPTH = 6
DEFAULT_MAX_PAYLOAD_KEYS = 50
DEFAULT_MAX_STRING_LENGTH = 2048
DEFAULT_MAX_QUEUE_EVENTS = 500
DEFAULT_MAX_QUEUE_BYTES = 2 * 1024 * 1024
DEFAULT_MAX_BATCH_SIZE = 20

REDACTION_MARKER = "[redacted]"
TRUNCATION_MARKER_KEY = "__trackmcp_truncated"
BASE64_SCRUB_THRESHOLD = 128

_DEFAULT_SENSITIVE_KEYS = {
    "password", "passwd", "secret", "token", "api_key", "apikey", "authorization",
    "cookie", "set_cookie", "access_token", "refresh_token", "private_key", "client_secret",
    "ssn", "credit_card", "card_number",
}
_BASE64_RE = re.compile(r"^[A-Za-z0-9+/_-]+={0,2}$")
_BEARER_RE = re.compile(r"\bbearer\s+[A-Za-z0-9._~+/-]+=*", re.IGNORECASE)
_CREDENTIAL_TEXT_RE = re.compile(r"(?:authorization|api[_-]?key|access[_-]?token|refresh[_-]?token|password|secret)\s*[:=]\s*\S+", re.IGNORECASE)


def normalize_key(key: str) -> str:
    key = re.sub(r"([a-z0-9])([A-Z])", r"\1_\2", key)
    return re.sub(r"[\s-]+", "_", key).lower()


def _marker(reason: str, original_type: str, **extra: Any) -> Dict[str, Any]:
    return {TRUNCATION_MARKER_KEY: True, "reason": reason, "original_type": original_type, **extra}


def _original_type(value: Any) -> str:
    if isinstance(value, list):
        return "array"
    if value is None:
        return "null"
    if isinstance(value, (bytes, bytearray, memoryview)):
        return "binary"
    if isinstance(value, dict):
        return "object"
    if isinstance(value, str):
        return "string"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, (int, float)):
        return "number"
    return type(value).__name__


def _canonical_json(value: Any) -> Optional[str]:
    try:
        return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False, allow_nan=False)
    except (TypeError, ValueError, OverflowError):
        return None


def payload_byte_length(value: Any) -> int:
    serialized = _canonical_json(value)
    return len(serialized.encode("utf-8")) if serialized is not None else 0


def _resource_marker(value: str) -> Optional[Dict[str, Any]]:
    match = re.match(r"^data:([^;,]+)?(?:;base64)?,", value, re.IGNORECASE)
    if match:
        return _marker(
            "binary",
            "string",
            media_type=match.group(1) or "application/octet-stream",
            original_bytes=max(0, int((len(value) - match.end()) * 0.75)),
        )
    if _CREDENTIAL_TEXT_RE.search(value):
        return _marker("credential", "string")
    if _BEARER_RE.search(value):
        return _marker("resource_uri", "string")
    try:
        parsed = urlparse(value)
        query_keys = {normalize_key(key) for key, _ in parse_qsl(parsed.query, keep_blank_values=True)}
        if parsed.username or parsed.password or query_keys & _DEFAULT_SENSITIVE_KEYS:
            return _marker("resource_uri", "string")
    except ValueError:
        pass
    if len(value) >= BASE64_SCRUB_THRESHOLD and _BASE64_RE.fullmatch(value):
        try:
            base64.b64decode(value + "=" * (-len(value) % 4), validate=True)
            return _marker("binary", "string", original_bytes=int(len(value) * 0.75))
        except (ValueError, binascii.Error):
            pass
    return None


def _sanitize_value(
    value: Any,
    depth: int,
    path: str,
    max_depth: int,
    max_keys: int,
    max_string_length: int,
    explicit_paths: Set[str],
    redact_keys: Set[str],
    seen: Set[int],
) -> Any:
    if path and path in explicit_paths:
        return REDACTION_MARKER
    if isinstance(value, str):
        resource = _resource_marker(value)
        if resource is not None:
            return resource
        return _marker("max_string_length", "string") if len(value) > max_string_length else value
    if value is None or isinstance(value, (int, float, bool)):
        return value
    if isinstance(value, (bytes, bytearray, memoryview)):
        return _marker("binary", "binary", original_bytes=len(value))
    if not isinstance(value, (dict, list)):
        return _marker("binary", type(value).__name__)
    identity = id(value)
    if identity in seen:
        return _marker("circular", _original_type(value))
    if depth >= max_depth:
        return _marker("max_payload_depth", _original_type(value))

    seen.add(identity)
    try:
        if isinstance(value, list):
            result = [
                _sanitize_value(item, depth + 1, f"{path}.{index}" if path else str(index), max_depth, max_keys, max_string_length, explicit_paths, redact_keys, seen)
                for index, item in enumerate(value[:max_keys])
            ]
            if len(value) > max_keys:
                result.append(_marker("max_payload_keys", "array"))
            return result
        result: Dict[str, Any] = {}
        keys = list(value.keys())
        for key in keys[:max_keys]:
            if not isinstance(key, str):
                continue
            child_path = f"{path}.{key}" if path else key
            if normalize_key(key) in _DEFAULT_SENSITIVE_KEYS or normalize_key(key) in redact_keys:
                result[key] = REDACTION_MARKER
            else:
                try:
                    result[key] = _sanitize_value(value[key], depth + 1, child_path, max_depth, max_keys, max_string_length, explicit_paths, redact_keys, seen)
                except Exception:
                    result[key] = _marker("binary", "unknown")
        if len(keys) > max_keys:
            result[TRUNCATION_MARKER_KEY] = _marker("max_payload_keys", "object")
        return result
    finally:
        seen.discard(identity)


def sanitize_payload(value: Any, *, mode: str, explicit_paths: Iterable[str] = (), redact_keys: Iterable[str] = (), max_payload_bytes: int = DEFAULT_MAX_PAYLOAD_BYTES, max_payload_depth: int = DEFAULT_MAX_PAYLOAD_DEPTH, max_payload_keys: int = DEFAULT_MAX_PAYLOAD_KEYS, max_string_length: int = DEFAULT_MAX_STRING_LENGTH) -> Any:
    if mode == "metadata" or value is None:
        return None if mode == "metadata" else value
    sanitized = _sanitize_value(
        value,
        0,
        "",
        max(0, int(max_payload_depth)),
        max(1, int(max_payload_keys)),
        max(1, int(max_string_length)),
        {".".join(part for part in path.split(".") if part) for path in explicit_paths if path},
        {normalize_key(key) for key in redact_keys},
        set(),
    )
    if payload_byte_length(sanitized) <= max(1, int(max_payload_bytes)):
        return sanitized
    budget_marker = _marker("max_payload_bytes", _original_type(value))
    return budget_marker if payload_byte_length(budget_marker) <= max(1, int(max_payload_bytes)) else {}

File: trackmcp-nextjs/packages/typescript-sdk/src/privacy.ts
export type PayloadMode = "metadata" | "redacted" | "full";

export const DEFAULT_MAX_PAYLOAD_BYTES = 32 * 1024;
export const DEFAULT_MAX_PAYLOAD_DEPTH = 6;
export const DEFAULT_MAX_PAYLOAD_KEYS = 50;
export const DEFAULT_MAX_STRING_LENGTH = 2048;
export const DEFAULT_MAX_QUEUE_EVENTS = 500;
export const DEFAULT_MAX_QUEUE_BYTES = 2 * 1024 * 1024;
export const DEFAULT_MAX_BATCH_SIZE = 20;

export const REDACTION_MARKER = "[redacted]";
export const TRUNCATION_MARKER_KEY = "__trackmcp_truncated";
const BASE64_SCRUB_THRESHOLD = 128;

const DEFAULT_SENSITIVE_KEYS = new Set([
  "password",
  "passwd",
  "secret",
  "token",
  "api_key",
  "apikey",
  "authorization",
  "cookie",
  "set_cookie",
  "access_token",
  "refresh_token",
  "private_key",
  "client_secret",
  "email",
  "e_mail",
  "ssn",
  "credit_card",
  "card_number",
]);

export type PayloadPrivacyOptions = {
  mode: PayloadMode;
  explicitPaths?: readonly string[];
  redactKeys?: readonly string[];
  maxPayloadBytes?: number;
  maxPayloadDepth?: number;
  maxPayloadKeys?: number;
  maxStringLength?: number;
};

type TruncationReason = "max_payload_depth" | "max_payload_keys" | "max_string_length" | "max_payload_bytes" | "binary" | "resource_uri" | "credential" | "circular";

function normalizedKey(key: string): string {
  return key
    .replace(/([a-z0-9])([A-Z])/g, "$1_$2")
    .replace(/[\s-]+/g, "_")
    .toLowerCase();
}

function marker(reason: TruncationReason, originalType: string, extra: Record<string, unknown> = {}): Record<string, unknown> {
  return { [TRUNCATION_MARKER_KEY]: true, reason, original_type: originalType, ...extra };
}

function originalType(value: unknown): string {
  if (Array.isArray(value)) return "array";
  if (value === null) return "null";
  if (value instanceof Uint8Array || value instanceof ArrayBuffer || ArrayBuffer.isView(value)) return "binary";
  return typeof value;
}

function byteLength(value: string): number {
  return new TextEncoder().encode(value).byteLength;
}

function safeJson(value: unknown): string | undefined {
  try {
    return JSON.stringify(value);
  } catch {
    return undefined;
  }
}

const SENSITIVE_TEXT_PATTERNS = [
  /\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/i,
  /\bbearer\s+[A-Za-z0-9._~+\/-]+=*/i,
  /\bbasic\s+[A-Za-z0-9+/=]{8,}/i,
  /\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b/,
  /-----BEGIN [A-Z ]*PRIVATE KEY-----/i,
  /\b(?:api[_-]?key|access[_-]?token|refresh[_-]?token|client[_-]?secret|password|secret|authorization)\s*[:=]\s*\S+/i,
];

function likelyBase64(value: string): boolean {
  return value.length >= BASE64_SCRUB_THRESHOLD && /^[A-Za-z0-9+/_-]+={0,2}$/.test(value);
}

function resourceMarker(value: string): Record<string, unknown> | undefined {
  const dataUri = value.match(/^data:([^;,]+)?(?:;base64)?,/i);
  if (dataUri) {
    return marker("binary", "string", {
      media_type: dataUri[1] || "application/octet-stream",
      original_bytes: Math.max(0, Math.floor((value.length - dataUri[0].length) * 0.75)),
    });
  }
  if (/(?:authorization|api[_-]?key|access[_-]?token|refresh[_-]?token|password|secret)\s*[:=]\s*\S+/i.test(value)) return marker("credential", "string");
  if (/\bbearer\s+[A-Za-z0-9._~+/-]+=*/i.test(value)) return marker("resource_uri", "string");
  if (SENSITIVE_TEXT_PATTERNS.some((pattern) => pattern.test(value))) return marker("credential", "string");
  try {
    const url = new URL(value);
    const hasCredentials = Boolean(url.username || url.password);
    const hasSensitiveQuery = [...url.searchParams.keys()].some((key) => DEFAULT_SENSITIVE_KEYS.has(normalizedKey(key)));
    if (hasCredentials || hasSensitiveQuery) return marker("resource_uri", "string");
  } catch {
    // Not a URL; it can still be an ordinary string or base64 value.
  }
  if (likelyBase64(value)) return marker("binary", "string", { original_bytes: Math.floor(value.length * 0.75) });
  return undefined;
}

function isSensitiveKey(key: string, extraKeys: ReadonlySet<string>): boolean {
  const normalized = normalizedKey(key);
  return DEFAULT_SENSITIVE_KEYS.has(normalized) || extraKeys.has(normalized);
}

function pathSet(paths: readonly string[]): ReadonlySet<string> {
  return new Set(paths.filter(Boolean).map((path) => path.split(".").filter(Boolean).join(".")));
}

function sanitizeValue(
  value: unknown,
  depth: number,
  path: string,
  options: Required<Pick<PayloadPrivacyOptions, "maxPayloadDepth" | "maxPayloadKeys" | "maxStringLength">>,
  explicitPaths: ReadonlySet<string>,
  extraKeys: ReadonlySet<string>,
  seen: WeakSet<object>,
): unknown {
  if (path && explicitPaths.has(path)) return REDACTION_MARKER;
  if (typeof value === "string") {
    const resource = resourceMarker(value);
    if (resource) return resource;
    return value.length > options.maxStringLength ? marker("max_string_length", "string") : value;
  }
  if (value === null || typeof value === "number" || typeof value === "boolean") return value;
  if (value === undefined) return undefined;
  if (value instanceof Uint8Array || value instanceof ArrayBuffer || ArrayBuffer.isView(value)) {
    return marker("binary", "binary", { original_bytes: value.byteLength });
  }
  if (typeof value !== "object") return marker("binary", typeof value);
  if (seen.has(value)) return marker("circular", originalType(value));
  if (depth >= options.maxPayloadDepth) return marker("max_payload_depth", originalType(value));

  seen.add(value);
  try {
    if (Array.isArray(value)) {
      const result = value.slice(0, options.maxPayloadKeys).map((item, index) => sanitizeValue(item, depth + 1, path ? `${path}.${index}` : String(index), options, explicitPaths, extraKeys, seen));
      if (value.length > options.maxPayloadKeys) result.push(marker("max_payload_keys", "array"));
      return result;
    }
    const record = value as Record<string, unknown>;
    const keys = Object.keys(record);
    const result: Record<string, unknown> = {};
    for (const key of keys.slice(0, options.maxPayloadKeys)) {
      const childPath = path ? `${path}.${key}` : key;
      try {
        if (isSensitiveKey(key, extraKeys)) result[key] = REDACTION_MARKER;
        else result[key] = sanitizeValue(record[key], depth + 1, childPath, options, explicitPaths, extraKeys, seen);
      } catch {
        result[key] = marker("binary", "unknown");
      }
    }
    if (keys.length > options.maxPayloadKeys) result[TRUNCATION_MARKER_KEY] = marker("max_payload_keys", "object");
    return result;
  } finally {
    seen.delete(value);
  }
}

export function payloadByteLength(value: unknown): number {
  const serialized = safeJson(value);
  return serialized === undefined ? 0 : byteLength(serialized);
}

export function sanitizePayload(value: unknown, options: PayloadPrivacyOptions): unknown {
  if (options.mode === "metadata" || value === undefined) return undefined;
  const limits = {
    maxPayloadDepth: Math.min(DEFAULT_MAX_PAYLOAD_DEPTH, Math.max(0, Math.floor(options.maxPayloadDepth ?? DEFAULT_MAX_PAYLOAD_DEPTH))),
    maxPayloadKeys: Math.min(DEFAULT_MAX_PAYLOAD_KEYS, Math.max(1, Math.floor(options.maxPayloadKeys ?? DEFAULT_MAX_PAYLOAD_KEYS))),
    maxStringLength: Math.min(DEFAULT_MAX_STRING_LENGTH, Math.max(1, Math.floor(options.maxStringLength ?? DEFAULT_MAX_STRING_LENGTH))),
  };
  const sanitized = sanitizeValue(value, 0, "", limits, pathSet(options.explicitPaths || []), new Set((options.redactKeys || []).map(normalizedKey)), new WeakSet());
  const maxBytes = Math.min(DEFAULT_MAX_PAYLOAD_BYTES, Math.max(1, Math.floor(options.maxPayloadBytes ?? DEFAULT_MAX_PAYLOAD_BYTES)));
  const serialized = safeJson(sanitized);
  if (serialized !== undefined && byteLength(serialized) <= maxBytes) return sanitized;
  const budgetMarker = marker("max_payload_bytes", originalType(value));
  return payloadByteLength(budgetMarker) <= maxBytes ? budgetMarker : {};
}
